R/Reporter.R

```diff
@@ -44,17 +45,25 @@ Reporter <- R6::R6Class( # nolint: object_name_linter.
     #' reporter <- Reporter$new()
     #' reporter$append_cards(list(card1, doc1))
     append_cards = function(cards) {
+      if (checkmate::test_multi_class(cards, classes = c("ReportDocument", "ReportCard"))) {
+        cards <- list(cards)
+      }
+
       checkmate::assert_list(cards, types = c("ReportCard", "ReportDocument"))
-      rcs <- which(vapply(cards, inherits, logical(1), "ReportCard"))
-      names(cards)[rcs] <- sapply(cards[rcs], function(card) card$get_name())
+      lapply(cards, function(x) checkmate::assert(private$check_append(x)))
+      new_cards <- cards
 
-      rds <- which(vapply(cards, inherits, logical(1), "ReportDocument"))
-      if (length(rds) && !is.null(self$get_template())) {
-        cards[rds] <- lapply(cards[rds], self$get_template())
+      rds <- vapply(new_cards, inherits, logical(1L), "ReportDocument")
+      if (!is.null(self$get_template())) {
+        new_cards[rds] <- lapply(new_cards[rds], self$get_template())
       }
-      cards <- mapply(private$update_attributes, card = cards, label = names(cards), SIMPLIFY = FALSE)
-      private$cards <- append(private$cards, cards)
-      shiny::isolate(private$reactive_add_card(length(private$cards)))
+
+      new_cards <- lapply(new_cards, private$update_attributes)
+
+      # Set up unique id for each card
+      names(new_cards) <- vapply(new_cards, metadata, character(1L), which = "id")
```

Comment:
Does "id" have to be set outside. Since it is random string we can set it in this method body instead.

Reply:
Done, and id was removed as `metadata`